api/proto/teleport/legacy/types/types.proto

```diff
@@ -7233,6 +7233,10 @@ message AWSOIDCIntegrationSpecV1 {
     (gogoproto.jsontag) = "issuer_s3_uri,omitempty",
     deprecated = true
   ];
+
+  // Audiences is used to record name of a plugin or discover services in Teleport
+  // that depends on this integration.
+  repeated string audiences = 3 [(gogoproto.jsontag) = "audiences,omitempty"];
```

Comment:
I think it should be more clear how we are planning on using the list of audiences on this comment.

What does it mean to have multiple values?
Does it mean we can use this integration for multiple services? As of right now, that would be `discovery.service` and `aws-identity-center`.

We'll need to change the IAM IdP configuration script to allow this list of audiences, so that the IAM Role's Trust Relationship, includes the `aws-identity-center`, `discovery.service` or both.

I guess this can come in a later PR.

Reply:
Updated commentary with more context on the usage of the `audiences ` https://github.com/gravitational/teleport/pull/47725/commits/f108acc487e30947f12d05da160c335bfa659813

It just explains the current intended usage, which is to prevent update/deletion of the integration used by `aws-identity`center` plugin.

Reply:
> We'll need to change the IAM IdP configuration script to allow this list of audiences, so that the IAM Role's Trust Relationship, includes the aws-identity-center, discovery.service or both.
> 
> I guess this can come in a later PR.

Yup. My 2c is that we should only focus on current use case and defer others until a plan is ready for them. Because most of the time, requirement for other service or use cases may come off completely different than today.

Reply:
We already have a new use case that we need to tackle: the DiscoveryService can use the Integration (and IAM Role) created for the AWS Identity Center.
It looks like the IAM Role for AWS IC will have a lot of capabilities and can escalate privileges.
So, this means that if a cluster has the AWS Identity Center integrated, then, any DiscoveryService can generate a token that allows the user to become AWS Admin.

I think we should add the following safeguard:
- new audience field in the integration
- the service GenerateAWSOIDCToken, when called from the DiscoveryService, can only do it for Integrations whose audience is empty (for back compat) or `discover.teleport`
- when signing the token, it will set the `aud` field to the integration's audience (or `discover.teleport` for empty audiences).


So, instead of allowing a list of audiences, we add a single one.
This also allows us to prevent deletion of the Integration

Note: I think we should have the same flow being used for preventing the deletion of an Integration based on the existence of something that depends on it. So, the External Audit Storage and AWS Identity Center should use the same logic.

Reply:
@marcoandredinis I think you got confused with the purpose of this PR and the use case to address OIDC integration access control. Related but completely different priority. 

> So, instead of allowing a list of audiences, we add a single one.
This also allows us to prevent deletion of the Integration

You literally suggested to support multiple audiences here https://github.com/gravitational/teleport/pull/47725#issuecomment-2426943853 :D

That is the reason I requested to discuss that OIDC integration access control thing separately. Trying to shoehorn different use case here is a sure recipe to make mistakes. Can we focus on the current use case, which is to prevent integration from deletion? 

I am updating this PR to use a single audience field, which likely will address both use case but lets take access control discussion out from this PR. If we find `audience` field will not completely address the access control, we can add a new field dedicated for that. Adding a new field is cheap but mudding purpose of that field can be expensive.

Reply:
https://github.com/gravitational/teleport/pull/47725/commits/7d990f0774fee54e05913bea388c40dc7e6423cd updates the `audiences` field to `audience` field which accepts a single preset value now.